File: Phan_B/bai41/bai_41.jl
function tich_lon_nhat(a)
    isempty(a) && return 0
    max_so_far = a[1]
    curr_max = a[1]
    curr_min = a[1]

    for i in 2:length(a)
        x = a[i]
        if x < 0
            curr_max, curr_min = curr_min, curr_max
        end
        curr_max = max(x, curr_max * x)
        curr_min = min(x, curr_min * x)
        max_so_far = max(max_so_far, curr_max)
    end
    return max_so_far
end

input_data = split(read(stdin, String))
if !isempty(input_data)
    n = parse(Int, input_data[1])
    a = [parse(Int, input_data[i]) for i in 2:(n+1)]
    println(tich_lon_nhat(a))
end

File: Phan_B/bai41/bai_41.py
import sys

def tich_lon_nhat(a):
    if not a:
        return 0
    max_so_far = a[0]
    curr_max = a[0]
    curr_min = a[0]

    for x in a[1:]:
        if x < 0:
            curr_max, curr_min = curr_min, curr_max
        curr_max = max(x, curr_max * x)
        curr_min = min(x, curr_min * x)
        max_so_far = max(max_so_far, curr_max)
    return max_so_far

def tich_lon_nhat_n2(a):
    n = len(a)
    max_prod = a[0]
    for i in range(n):
        p = 1
        for j in range(i, n):
            p *= a[j]
            max_prod = max(max_prod, p)
    return max_prod

if __name__ == "__main__":
    input_data = sys.stdin.read().split()
    if input_data:
        n = int(input_data[0])
        a = [int(x) for x in input_data[1:n+1]]
        print(tich_lon_nhat(a))